### ISSUE TITLE
MPS: catastrophically wrong gradients in backward pass (>32K elements)

### ISSUE THREAD

**Issue by erozon** [pytorch/pytorch#177116]
### 🐛 Describe the bug

The MPS backend produces catastrophically wrong gradients (1,000x–100,000x too large) during `loss.backward()` when:

1. A prior MPS forward+backward pass has occurred in the same process at a **different batch size**, and
2. The total number of elements (batch_size × seq_len) exceeds **~32,768 (2^15)**.

The forward pass is always correct — loss values match CPU within float32 tolerance. Only the backward pass is affected. Without the prior MPS operation at a different shape, gradients are correct. In some quick experimentation, `torch.mps.empty_cache()` between calls appeared to reduce the failure rate, suggesting MPS buffer pool corruption when tensor shapes change between backward passes.

Possibly related to #116769 (int16 overflow in Metal matmul shaders, fixed for `torch.mm`/`torch.bmm` via tiling in PR #117549 but not all code paths), #122045 (`F.linear` wrong results for large inputs), #117826 (incorrect einsum gradient on MPS).

### Minimal reproduction

The model is `Embedding → 2 residual Linear blocks → output Linear`, trained with `CrossEntropyLoss`. Step 1 primes MPS at a different batch size; step 2 tests at batch=8194 (32,776 total elements > 32,768).

```python
import torch
import torch.nn as nn


class ResidualModel(nn.Module):
    def __init__(self, vocab_size=122, d_model=128):
        super().__init__()
        self.vocab_size = vocab_size
        self.emb = nn.Embedding(vocab_size, d_model)
        self.fc1 = nn.Linear(d_model, d_model, bias=False)
        self.fc2 = nn.Linear(d_model, d_model, bias=False)
        self.out = nn.Linear(d_model, vocab_size, bias=False)

    def forward(self, x):
        h = self.emb(x)
        h = h + self.fc1(h)  # residual 1
        h = h + self.fc2(h)  # residual 2
        return self.out(h)


V = 122
seq_len = 4
criterion = nn.CrossEntropyLoss()

# Step 1: prime MPS with a forward+backward at a DIFFERENT batch size.
# Without this step, MPS produces correct gradients.
torch.manual_seed(0)
prime_x = torch.randint(0, V, (4096, seq_len)).to("mps")
prime_y = torch.randint(0, V, (4096, seq_len)).view(-1).to("mps")
torch.manual_seed(0)
prime_m = ResidualModel().to("mps")
prime_loss = criterion(prime_m(prime_x).view(-1, V), prime_y)
prime_loss.backward()
del prime_m, prime_loss, prime_x, prime_y

# Step 2: test at batch=8194 — total elements = 8194*4 = 32776 > 32768
batch_size = 8194
torch.manual_seed(0)
x = torch.randint(0, V, (batch_size, seq_len))
y = torch.randint(0, V, (batch_size, seq_len))

for device in ["cpu", "mps"]:
    results = []
    for trial in range(5):
        torch.manual_seed(0)
        model = ResidualModel().to(device)
        logits = model(x.to(device))
        loss = criterion(logits.view(-1, V), y.view(-1).to(device))
        loss.backward()
        gnorm = sum(
            p.grad.norm().item() ** 2 for p in model.parameters()
        ) ** 0.5
        results.append((loss.item(), gnorm))

    print(f"\n{device.upper()}:")
    for i, (l, g) in enumerate(results):
        print(f"  trial {i}: loss={l:.6f}  grad_norm={g:.4f}")
```

### Expected output

CPU and MPS should produce the same loss and gradient norm across all trials.

### Actual output

```
CPU:
  trial 0: loss=5.089585  grad_norm=0.2398
  trial 1: loss=5.089585  grad_norm=0.2398
  trial 2: loss=5.089585  grad_norm=0.2398
  trial 3: loss=5.089585  grad_norm=0.2398
  trial 4: loss=5.089585  grad_norm=0.2398

MPS:
  trial 0: loss=5.089585  grad_norm=0.2414       ← within float32 tolerance
  trial 1: loss=5.089585  grad_norm=3529.6575     ← 14,719x too large
  trial 2: loss=5.089585  grad_norm=16290.2734    ← 67,932x too large
  trial 3: loss=5.089585  grad_norm=1996.1491     ← 8,324x too large
  trial 4: loss=5.089585  grad_norm=16290.2734    ← 67,932x too large
```

Loss is identical. Gradient norms are wrong by 1,000x–68,000x on MPS. Trial 0 is correct; trials 1+ are catastrophically wrong (each backward appears to corrupt buffer state for subsequent trials). Without step 1, all MPS trials match CPU.

We posted additional analysis (sweep data across batch sizes and sequence lengths, `empty_cache` observations, and diagnostic experiments pointing to MPS buffer pool reuse) as a follow-up comment below.

### Versions

PyTorch version: 2.10.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 26.2 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.3.9.4)
CMake version: version 4.2.0
Libc version: N/A

Python version: 3.12.8 (main, Aug  4 2025, 07:08:45) [Clang 15.0.0 (clang-1500.3.9.4)] (64-bit runtime)
Python platform: macOS-26.2-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Apple M2

Versions of relevant libraries:
[pip3] numpy==2.3.5
[pip3] torch==2.10.0
[conda] Could not collect

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93 @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk

**Comment by Isalia20** [comment]
@erozon Can you confirm that it works on the latest torch version?

**Comment by erozon** [comment]
Tested on torch **2.11.0** (M2, macOS 26.2) with the same script — bug still reproduces. Sample of the diff vs 2.10.0:

| size | 2.10.0 max ratio | 2.11.0 max ratio |
|---|---:|---:|
| 8194×4 | 20,402x | 70,141x |
| 4098×8 | 74,701x | 120,203x |
| 2049×16 | 74,701x | 120,203x |
| 16385×4 | 27,043x | 1.0x ✓ |

The 16385×4 case is clean on 2.11.0, but the threshold (batch × seq > 32,768), the priming requirement, and forward-pass correctness are all unchanged, and error magnitudes are larger on 2.11.0 for several sizes.

@Isalia20 this could be a chip/macOS version issue... Wondering whether this is M2- or macOS 26.x-specific. One note in case it helps: a single batch size in a fresh process won't trigger it, as the sweep itself does the priming.

<details>
<summary>Full output: torch 2.10.0</summary>

```
PyTorch 2.10.0
MPS available: True

Model: Embedding(122,128) -> 2x residual Linear(128,128) -> Linear(128,122)
Loss:  CrossEntropyLoss
Each row: 20 independent MPS trials with the same seed

--- Batch size sweep (seq_len=4) ---
    Threshold expected near batch*4 > 32768, i.e., batch > 8192

  batch=  4096  seq= 4  tokens= 16384  loss= OK  bad= 0/20  max_ratio=       1.0x
  batch=  8000  seq= 4  tokens= 32000  loss= OK  bad= 0/20  max_ratio=       1.0x
  batch=  8192  seq= 4  tokens= 32768  loss= OK  bad= 0/20  max_ratio=       1.0x
  batch=  8193  seq= 4  tokens= 32772  loss= OK  bad= 0/20  max_ratio=       1.0x
  batch=  8194  seq= 4  tokens= 32776  loss= OK  bad=19/20  max_ratio=   20402.5x  *** BUG
  batch=  8200  seq= 4  tokens= 32800  loss= OK  bad=19/20  max_ratio=   17661.2x  *** BUG
  batch= 10000  seq= 4  tokens= 40000  loss= OK  bad=18/20  max_ratio=    8725.0x  *** BUG
  batch= 16384  seq= 4  tokens= 65536  loss= OK  bad= 0/20  max_ratio=       1.0x
  batch= 16385  seq= 4  tokens= 65540  loss= OK  bad=19/20  max_ratio=   27042.6x  *** BUG

--- Sequence length sweep (confirming threshold is batch*seq > ~32768) ---

  batch=  8192  seq= 4  tokens= 32768  loss= OK  bad= 0/20  max_ratio=       1.0x
  batch=  8193  seq= 4  tokens= 32772  loss= OK  bad= 0/20  max_ratio=       1.0x
  batch=  8194  seq= 4  tokens= 32776  loss= OK  bad=20/20  max_ratio=   20141.8x  *** BUG

  batch=  4096  seq= 8  tokens= 32768  loss= OK  bad= 0/20  max_ratio=       1.0x
  batch=  4097  seq= 8  tokens= 32776  loss= OK  bad=20/20  max_ratio=   20141.8x  *** BUG
  batch=  4098  seq= 8  tokens= 32784  loss= OK  bad=20/20  max_ratio=   74701.1x  *** BUG

  batch=  2048  seq=16  tokens= 32768  loss= OK  bad= 0/20  max_ratio=       1.0x
  batch=  2049  seq=16  tokens= 32784  loss= OK  bad=20/20  max_ratio=   74701.1x  *** BUG
  batch=  2050  seq=16  tokens= 32800  loss= OK  bad=20/20  max_ratio=   39915.2x  *** BUG
```

</details>

<details>
<summary>Full output: torch 2.11.0</summary>

```
PyTorch 2.11.0
MPS available: True

Model: Embedding(122,128) -> 2x residual Linear(128,128) -> Linear(128,122)
Loss:  CrossEntropyLoss
Each row: 20 independent MPS trials with the same seed

--- Batch size sweep (seq_len=4) ---
    Threshold expected near batch*4 > 32768, i.e., batch > 8192

  batch=  4096  seq= 4  tokens= 16384  loss= OK  bad= 0/20  max_ratio=       1.0x
  batch=  8000  seq= 4  tokens= 32000  loss= OK  bad= 0/20  max_ratio=       1.0x
  batch=  8192  seq= 4  tokens= 32768  loss= OK  bad= 0/20  max_ratio=       1.0x
  batch=  8193  seq= 4  tokens= 32772  loss= OK  bad= 0/20  max_ratio=       1.0x
  batch=  8194  seq= 4  tokens= 32776  loss= OK  bad=19/20  max_ratio=   70141.8x  *** BUG
  batch=  8200  seq= 4  tokens= 32800  loss= OK  bad=19/20  max_ratio=   17661.2x  *** BUG
  batch= 10000  seq= 4  tokens= 40000  loss= OK  bad=18/20  max_ratio=    8725.0x  *** BUG
  batch= 16384  seq= 4  tokens= 65536  loss= OK  bad= 0/20  max_ratio=       1.0x
  batch= 16385  seq= 4  tokens= 65540  loss= OK  bad= 0/20  max_ratio=       1.0x

--- Sequence length sweep (confirming threshold is batch*seq > ~32768) ---

  batch=  8192  seq= 4  tokens= 32768  loss= OK  bad= 0/20  max_ratio=       1.0x
  batch=  8193  seq= 4  tokens= 32772  loss= OK  bad= 0/20  max_ratio=       1.0x
  batch=  8194  seq= 4  tokens= 32776  loss= OK  bad=20/20  max_ratio=   70111.8x  *** BUG

  batch=  4096  seq= 8  tokens= 32768  loss= OK  bad= 0/20  max_ratio=       1.0x
  batch=  4097  seq= 8  tokens= 32776  loss= OK  bad=20/20  max_ratio=   70111.8x  *** BUG
  batch=  4098  seq= 8  tokens= 32784  loss= OK  bad=20/20  max_ratio=  120202.8x  *** BUG

  batch=  2048  seq=16  tokens= 32768  loss= OK  bad= 0/20  max_ratio=       1.0x
  batch=  2049  seq=16  tokens= 32784  loss= OK  bad=20/20  max_ratio=  120202.8x  *** BUG
  batch=  2050  seq=16  tokens= 32800  loss= OK  bad=20/20  max_ratio=   49801.6x  *** BUG
```

</details>

**Comment by Isalia20** [comment]
I have M2, I'll try to reproduce it tomorrow

**Comment by Isalia20** [comment]
Reproduced it on M2, need to find what's causing this, M2+ macs are fine (couldn't check on M1 due to not having the hardware)

**Comment by Lemniscate-world** [comment]
Hi, I investigated this issue with [NeuralDBG](https://github.com/LambdaSection/NeuralDBG), a causal diagnostic tool for PyTorch training.

### NeuralDBG Detection

NeuralDBG installs backward hooks that track per-layer gradient norms and detect health transitions. On this bug, it detects:

- **gradient_health_transition** event: gradient norms explode from ~0.24 to ~3,500 between consecutive trials
- **Hypothesis**: "Gradient explosion detected on Linear_0 / Linear_1" with confidence 1.0
- **Localization**: Identifies the specific layers where gradient norms changed

This confirms the bug is a **silent backward pass corruption** -- forward pass produces correct loss, but gradients are wrong by 1,000x-68,000x.

### Key Insight

This is exactly the class of bug NeuralDBG is designed to detect: the loss looks fine (you'd never know from W&B/TensorBoard), but the gradients are catastrophically wrong. The model appears to train but produces no learning (loss stuck at ~0.55 for 80 epochs in the VAE case).

### Reproduction Summary

Using the upstream repro (ResidualModel with MPS, batch_size x seq_len > 32,768):

```
Trial 0: loss=5.089585  grad_norm=0.2414      <- correct
Trial 1: loss=5.089585  grad_norm=3529.6575   <- 14,719x too large
Trial 2: loss=5.089585  grad_norm=16290.2734  <- 67,932x too large
```

Loss is identical. Only gradient norms are corrupted.

### Suggested Investigation

The correlation with MPS buffer pool reuse (`torch.mps.empty_cache()` reduces failure rate) suggests the bug is in the MPS backend's cached buffer management during backward passes when tensor shapes change between trials.

### Environment

- PyTorch: 2.10.0
- macOS 26.2 (Darwin 25.2.0)
- Apple Silicon (M-series)

**Comment by Lemniscate-world** [comment]
Requesting this issue be reviewed for **actionable** status:

**Bug**: MPS backend produces catastrophically wrong gradients (100x-100,000x too large) for certain operations on Apple Silicon.

**Impact**: Anyone training on MPS Macs gets silently corrupted models. The gradients are wrong but finite, so no NaN alert is raised — the model just converges to incorrect weights.

**Detection**: We've developed gradient health validation tests that would catch this class of bug across any backend (CPU, CUDA, MPS). The tests establish baseline expectations:
- Gradients must be finite after normal backward pass
- Gradients must not explode on scaled inputs
- Gradients must scale linearly with loss scaling

These tests are backend-agnostic and would have caught #177116 immediately on MPS hardware. PR #188923 has the test file ready.

Could a maintainer advise on current status of the MPS fix and whether gradient health tests would be welcome? Thank you!
